File: G_Projects/I_Vending_Machine/VendingMachineStates/ConcreteStates/IdleState.java
package G_Projects.I_Vending_Machine.VendingMachineStates.ConcreteStates;

import G_Projects.I_Vending_Machine.CommonEnums.Coin;
import G_Projects.I_Vending_Machine.VendingMachineStates.VendingMachineContext;
import G_Projects.I_Vending_Machine.VendingMachineStates.VendingMachineState;

public class IdleState implements VendingMachineState {

    @Override
    public String getStateName(){
        return "Idle State";
    }

    @Override
    public void insertCoin(VendingMachineContext context, Coin coin){
        context.addCoin(coin);
        context.setState(new HasMoneyState());
    }

    @Override
    public void selectItem(VendingMachineContext context, int itemCode){
        throw new IllegalStateException("Insert money first");
    }

    @Override
    public void dispenseItem(VendingMachineContext context){
        throw new IllegalStateException("Nothing to dispense in Idle State");
    }
}

File: Python/A_Project/E_Vending_Machine/VendingMachineStates/ConcreteStates/idle_state.py
from VendingMachineStates.vending_machine_state import VendingMachineState


class IdleState(VendingMachineState):
    def get_state_name(self) -> str:
        return "Idle State"

    def insert_coin(self, context, coin):
        # To handle circular import issue, we import IdleState here instead of at the top
        from VendingMachineStates.ConcreteStates.has_money_state import HasMoneyState

        context.add_coin(coin)
        context.set_state(HasMoneyState())

    def select_item(self, context, item_code: int):
        raise RuntimeError("Insert money first")

    def dispense_item(self, context):
        raise RuntimeError("Nothing to dispense in Idle State")
